File: Stack/C++/Implementation/Array/Stack.cpp
//
// Created by vivekdagar on 19/10/23.
//

#include <iostream>
#include "Stack.h"

Stack::Stack(unsigned short size) : size(size) {
    array.resize(this->size);
    top = -1;
}

void Stack::push(int x) {           // Insert an element into the stack (top)
    if (isFull()) {
        std::cerr << "Stack overflow!\n";
        return;
    }

    this->top++;
    this->array.at(top) = x;
}

int Stack::pop() {          // Remove an element from stack (top)
    int x{-1};

    if (isEmpty()) {
        std::cerr << "Stack underflow!\n";
        return x;
    }

    x = this->array[top];
    array.pop_back();
    this->top--;
    return x;
}

std::ostream &operator<<(std::ostream &os, const Stack &stack) {            // Printing the stack using cout
    for (int i{stack.top}; i >= 0; i--)
        os << stack.array.at(i) << std::endl;
    os << std::endl;
    return os;
}

Stack::~Stack() {
    array.clear();
}

File: Stack/Java/Stack/src/implementation/array/Stack.java
package implementation.array;

public class Stack {
    private final int size;
    private int top;
    private final int[] array;

    public Stack(int size) {
        this.size = size;
        this.array = new int[size];
        this.top = -1;
    }

    public boolean isFull() {
        return this.top == this.size - 1;
    }

    public void push(int x) {
        if (isFull()) {
            System.err.println("Stack overflow!");
            return;
        }
        this.top++;
        this.array[top] = x;
    }

    public boolean isEmpty() {
        return this.top == -1;
    }

    public int pop() {
        if (isEmpty()) {
            System.err.println("Stack underflow!");
            return -1;
        }
        int x = this.array[top];
        this.top--;
        return x;
    }

    public int peek() {
        return isEmpty() ? -1 : this.array[top];
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = this.top; i >= 0; i--) {
            sb.append(this.array[i]).append("\n");
        }
        sb.append("\n");
        return sb.toString();
    }
}

